src/Http/Routing/src/EndpointMiddlewareEventSource.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics.Tracing;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal class EndpointMiddlewareEventSource : EventSource
+    {
+        public static readonly EndpointMiddlewareEventSource Log = new EndpointMiddlewareEventSource();
+
+        private readonly ConcurrentDictionary<string, EndpointCounter> _endpointCounters = new ConcurrentDictionary<string, EndpointCounter>();
+
+        internal EndpointMiddlewareEventSource()
+            : base("Microsoft.AspNetCore.Routing")
+        {
+
+        }
+
+        [Event(1, Level = EventLevel.Informational)]
+        public void ExecutingEndpoint(string endpoint)
+        {
+            var endpointCounter = _endpointCounters.GetOrAdd(endpoint, key => new EndpointCounter(key, this));
```

Comment:
Looking up a dictionary multiple times (executing, executed, failed) for every call would have a performance impact. This code is on a very hot path.

I don't know the best way to mitigate but EventSource.OnEventCommand provides a hook to decide to enable counters. We should only be initializing counters when needed.

Reply:
https://github.com/dotnet/aspnetcore/blob/5a0c097ad4b0aef756cd92f2d593cda3fa223748/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs#L314-L319

Reply:
The problem is that even if the event source is disabled, we would still need to do the dictionary lookup to ensure `EndpointCounter._totalRequests/_currentRequests/_failedRequests` are all accurate when the event source becomes enabled.

We can still override OnEventCommand to lazily initialize the PollingCounters, but we'd have to iterate `_endpointCounters` dictionary when the event source becomes enabled to do so.

Reply:
Ah, right. We could add counters as metadata or a property on the endpoint. However endpoints are intended to be immutable and Ryan would murder us all 😄

Reply:
I was thinking the same thing. Problem with a property is that Endpoint is defined in another assembly and InternalsVisibleTo is bad. Putting it in Metadata doesn't work because it *is* immutable and Ryan would murder us. Retrieving the EndpointCounter from Metadata would involve a dictionary lookup anyway.

Reply:
Another option is to have the routing matches store the counter, and matching an endpoint would return endpoint+counter tulple.

Would need to store it somewhere until endpoint invocation.

Reply:
Or have a setting enabling per-endpoint counters. By default they are off, and if you value the counters over the the perf loss then you'd set a flag on `RouteOptions`.

Reply:
> Another option is to have the routing matcher store the counter, and matching an endpoint would return endpoint+counter tulple.

That seems interesting. Would we create a new internal feature interface just for storing the counter?

Reply:
Yes.

There would be some other challenges, like endpoints can change. For example, dropping a new Razor file into a directory would change that endpoint data source and the matcher would recompute the DFA graph. When that happens we'd need to keep the endpoint+counter instances together.

Reply:
What should happen if someone wants to [`SetEndpoint`](https://github.com/dotnet/aspnetcore/blob/f45250ce5e423dad083fe99967a4d65782c5301e/src/Http/Http.Abstractions/src/Routing/EndpointHttpContextExtensions.cs#L34) without Routing Middleware. Do they need to provide the counter data as well?

Correct me if I'm wrong,  but Routing matcher only gives `RouteEndpoint` not `Endpoint`, and If we are going with  getting counters from Routing matcher, can we instead add a property to `RouteEndpoint` and since it's in the same assembly as there's no need to `InternalVisibleTo`.

Reply:
Routing matching usual matches `RouteEndpoint` but can return `Endpoint` from a policy.

That's a good point about SetEndpoint. Matching is the primary way of setting endpoints, but not the only way. Matching providing the counter type wouldn't work.

Reply:
@davidfowl Thoughts?

Reply:
@noahfalk This is an example of where we want dimensions for metrics to avoid the dictionary workaround.